src/utils/computeReClammInitAmounts.ts

```diff
@@ -0,0 +1,109 @@
+import { computeInitialBalanceRatio } from '@balancer-labs/balancer-maths';
+
+import { InputAmount } from '@/types';
+import { MinimalTokenWithRate } from '@/data/types';
+import { getSortedTokens } from '@/entities/utils/getSortedTokens';
+import { Token } from '@/entities/token';
+import { TokenAmount } from '@/entities/tokenAmount';
+import { TokenAmountWithRate } from '@/entities/tokenAmountWithRate';
+import { isSameAddress } from './helpers';
+import { MathSol } from './math';
+
+export type ComputeReClammInitAmountInput = {
+    chainId: number;
+    tokens: MinimalTokenWithRate[];
+    referenceAmountIn: InputAmount;
+    initialMinPrice: bigint;
+    initialMaxPrice: bigint;
+    initialTargetPrice: bigint;
+};
+
+export async function computeReClammInitAmounts({
+    chainId,
+    tokens,
+    referenceAmountIn,
+    initialMinPrice,
+    initialMaxPrice,
+    initialTargetPrice,
+}: ComputeReClammInitAmountInput): Promise<InputAmount[]> {
+    // sort the user provided tokens just in case
+    const sortedTokens = getSortedTokens(tokens, chainId);
+    const givenTokenIndex = sortedTokens.findIndex((t) =>
+        isSameAddress(t.address, referenceAmountIn.address),
+    );
+    const calculatedTokenIndex = givenTokenIndex === 0 ? 1 : 0;
+
+    const proportion = computeInitialBalanceRatio(
+        initialMinPrice,
+        initialMaxPrice,
+        initialTargetPrice,
+    );
+
+    const referenceToken = new Token(
+        chainId,
+        referenceAmountIn.address,
+        referenceAmountIn.decimals,
+        undefined, // symbol
+        undefined, // name
+        undefined, // wrapped
+        sortedTokens[givenTokenIndex]?.rate,
+    );
+    const computedToken = new Token(
+        chainId,
+        sortedTokens[calculatedTokenIndex].address,
+        sortedTokens[calculatedTokenIndex].decimals,
+        undefined, // symbol
+        undefined, // name
+        undefined, // wrapped
+        sortedTokens[calculatedTokenIndex]?.rate,
+    );
+
+    // Scale given amount to 18 decimals before doing math
+    const referenceTokenAmount = referenceToken.rate
+        ? TokenAmountWithRate.fromRawAmountWithRate(
+              referenceToken,
+              referenceAmountIn.rawAmount,
+              referenceToken.rate,
```

Comment:
I'm trying to wrap my head around the idea of adding rate to `Token` 🤔
I was expecting it would be only added to `TokenAmountWithRate` because that's where we keep concepts like amounts, decimals and rates.
Unless I'm missing something, this is the only place you're using it, right? If yes, you should be able to remove it from here without losing much.
You could replace with something like this:
```suggestion
              sortedTokens[givenTokenIndex]?.rate,
```

What do you think?

Reply:
Ya idk why I thought that was a good idea yesterday :thinking: 

I agree and like your suggestion much better

Reply:
oh wait I remember now . i did it because `getSortedTokens` returns type `Token` w/o rate

https://github.com/balancer/b-sdk/blob/3aff1f0640330bee66f339541840d9a1882d7194/src/entities/utils/getSortedTokens.ts#L7-L10

but there are still other ways to solve w/o changing `Token` i will explore

Reply:
the tricky part is a smooth way to ask for the rate of the computed token as part of input to `computeReClammInitAmounts` :thinking:

Reply:
@brunoguerios  how would you feel about using the pool address to on chain fetch the token config to get rate provider address then fetch the rate for each token?

Reply:
we could also fetch the initial params using the pool address so user only has to provide chainId, pool addy, and reference token / amount in

https://github.com/balancer/reclamm/blob/44b1d60324746aaf9c0ef3cbac6ffa763f134531/contracts/interfaces/IReClammPool.sol#L401

Reply:
Taking a step back, why aren't we querying that onchain? ReClamm has a function to `computeInicialBalances`, right?